File: 1484-linked-list-in-binary-tree/1484-linked-list-in-binary-tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubPath(ListNode head, TreeNode root) {
        if(root==null) return false;
        boolean ans = dfs(root, head);
        ans |= isSubPath(head, root.left);
        ans |= isSubPath(head, root.right);
        return ans;
    }
    boolean dfs(TreeNode node, ListNode head){
        if(head == null) return true;
        if(node == null) return false;
        boolean ans = false;
        if(node.val == head.val){
            ans |= dfs(node.left, head.next);
            ans |= dfs(node.right, head.next);
        }
        return ans;
    }
}

File: 1484-linked-list-in-binary-tree/1484-linked-list-in-binary-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        if(!root) return false;
        bool ans = dfs(root, head);
        ans |= isSubPath(head, root->left);
        ans |= isSubPath(head, root->right);
        return ans;
    }
    bool dfs(TreeNode *node, ListNode *head){
        if(!head) return true;
        if(!node) return false;
        bool ans = false;
        if(node->val == head->val){
            ans |= dfs(node->left, head->next);
            ans |= dfs(node->right, head->next);
        }
        return ans;
    }
};